packages/lodestar/src/network/gossip/gossipsub.ts

```diff
@@ -141,24 +144,46 @@ export class LodestarGossipsub extends Gossipsub {
    */
   public _emitMessage(message: InMessage): void {
     message.topicIDs.forEach((topic) => {
-      const transformedObj = this.transformedObjects.get(getMessageId(message));
+      const transformedObj = this.transformedObjects.get(msgIdToString(this.getMsgId(message)));
       if (transformedObj && transformedObj.object) {
         super.emit(topic, transformedObj.object);
       }
     });
   }
 
   /**
-   * Override https://github.com/ChainSafe/js-libp2p-gossipsub/blob/v0.6.3/ts/index.ts#L1034
+   * Override default `_buildMessage` to snappy-compress the data
    */
-  public _publish(message: InMessage): Promise<void> {
-    assert.true(message.topicIDs && message.topicIDs.length === 1, "lodestar only supports 1 topic per message");
-    assert.true(!!message.data, "message to publish should have data");
-    const encoding = getTopicEncoding(message.topicIDs[0]);
+  public _buildMessage(msg: InMessage): Promise<InMessage> {
+    const encoding = getTopicEncoding(msg.topicIDs[0]);
     if (encoding === GossipEncoding.SSZ_SNAPPY) {
-      message.data = compress<Uint8Array>(message.data!);
+      msg.data = compress(msg.data!);
+    }
+    return super._buildMessage(msg);
+  }
+
+  /**
+   * Override default `getMsgId` function to cache the message-id
+   */
+  public getMsgId(msg: ILodestarGossipMessage): Uint8Array {
+    if (!msg.msgId) {
+      msg.msgId = this.computeMsgId(msg);
+    }
+    return msg.msgId;
+  }
+
+  private computeMsgId(msg: InMessage): Uint8Array {
+    const encoding = getTopicEncoding(msg.topicIDs[0]);
+    const data = msg.data!;
+    if (encoding === GossipEncoding.SSZ_SNAPPY) {
+      try {
+        const uncompressed = uncompress(data);
+        return hash(Buffer.concat([MESSAGE_DOMAIN_VALID_SNAPPY, uncompressed])).slice(0, 20);
+      } catch (e) {
+        return hash(Buffer.concat([MESSAGE_DOMAIN_INVALID_SNAPPY, data])).slice(0, 20);
+      }
     }
-    return super._publish(message);
+    return hash(data).slice(0, 20);
```

Comment:
make this `20` a variable such as `GOSSIP_SUB_MSG_ID_LENGTH`

Reply:
Would make sense to add it to packages/lodestar/src/network/gossip/constants.ts ?